File: java/src/main/java/com/example/aggregation/Task2.java
package com.example.aggregation;

import com.mongodb.client.MongoClient;
import com.mongodb.client.MongoCollection;
import com.mongodb.client.MongoDatabase;
import org.bson.Document;

/*
 Task 2: Find acupuncture businesses with wait-related complaints
 ================================================================
 This file connects to the 'yelp.business' collection.

 Your task is to find acupuncture businesses that reviews related to wait times.
  1. Find open businesses (`is_open` == true) with 'Acupuncture' in `categories`
  2. Join the `business` and `review` collections by their `business_id` fields using `$lookup`
  3. Unwind the reviews array to process each review individually
  4. Filter for review `text` mentioning wait-related complaints ("wait", "waited", "waiting", "slow service") using a regex match
  5. Group filtered results by:
        - `businessName`
        - `city`
        - `waitReviewCount` (count of reviews per business)
  6. Return only the unique `businessName`, `city`, and`waitReviewCount` fields
  7. Sort by `waitReviewCount` in descending order

 See:
 - Aggregation: https://www.mongodb.com/docs/drivers/java/sync/current/aggregation/
 - Aggregates Builders: https://www.mongodb.com/docs/drivers/java/sync/current/builders/aggregates/
*/

public class Task2 {

    public static void main(String[] args) {
        try (MongoClient mongoClient = MongoClientProvider.getClient()) {

            MongoDatabase database = mongoClient.getDatabase("yelp");
            MongoCollection<Document> businessColl = database.getCollection("business");

//            var pipeline =
//                    Arrays.asList(

            // TODO: Add your aggregation logic here, then compile and run: `mvn exec:java -P2`

//            );

//            AggregateIterable<Document> results = businessColl.aggregate(pipeline);
//            for (Document doc : results) {
//                System.out.println(doc.toJson());
//            }
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
        System.exit(0);
    }
}

File: python/tasks/task2.py
# tasks/task2.py

import json
from mongo_client import get_client
from config import DATABASE_NAME

def run():
    """
    Task 2: Find acupuncture businesses with wait-related complaints
    1. Filter for open businesses with 'Acupuncture' in categories
    2. Join to the `review` collection on business_id
    3. Unwind the joined reviews array
    4. Filter for reviews whose text matches “wait(ed|ing)?” or “slow service” (case-insensitive)
    5. Group by business name and city, counting those reviews
    6. Project businessName, city, and waitReviewCount; sort descending
    """
    client = get_client()
    db = client[DATABASE_NAME]

    pipeline = [
        {"$match": {
            "is_open": True,
            "categories": "Acupuncture"
        }},
        {"$lookup": {
            "from": "review",
            "localField": "business_id",
            "foreignField": "business_id",
            "as": "reviews"
        }},
        {"$unwind": "$reviews"},
        {"$match": {
            "reviews.text": {
                "$regex": r"(wait(ed|ing)?|slow service)",
                "$options": "i"
            }
        }},
        {"$group": {
            "_id": "$name",
            "city": {"$first": "$city"},
            "waitReviewCount": {"$sum": 1}
        }},
        {"$project": {
            "_id": 0,
            "businessName": "$_id",
            "city": 1,
            "waitReviewCount": 1
        }},
        {"$sort": {"waitReviewCount": -1}}
    ]

    results = list(db.business.aggregate(pipeline))
    print(json.dumps(results, indent=2))
    client.close()

if __name__ == "__main__":
    run()
